### PR TITLE
webgpu: matmul split k and reduce

### PR DESCRIPTION
**Pull request by qjia7**
This  PR is targeted for small M and N, but large K.
 
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6285)

### CODE REVIEW

**Comment by qjia7** [comment]
Before
<html>
<body>


Conv2D | 1.78 | input0: 4D[1,4,6,2048]input1: 4D[1,1,2048,34] | 1,4,6,34 | MatMulPackedProgram: 1.783305
-- | -- | -- | -- | --
Conv2D | 1.76 | input0: 4D[1,4,6,2048]input1: 4D[1,1,2048,34] | 1,4,6,34 | MatMulPackedProgram: 1.762722


</body>
</html>

After
<html>
<body>


Conv2D | 0.32 | input0: 4D[1,4,6,2048]input1: 4D[1,1,2048,34] | 1,4,6,34 | FillProgram: 0.013333, MatMulSplitKProgram: 0.302996
-- | -- | -- | -- | --
Conv2D | 0.32 | input0: 4D[1,4,6,2048]input1: 4D[1,1,2048,34] | 1,4,6,34 | FillProgram: 0.013666, MatMulSplitKProgram: 0.302412


</body>
</html>

**Review by gyagp — APPROVED** [review]
LGTM

**Review by Linchenn — APPROVED** [review]
LGTM, thanks!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6285#-:-N5bIJSHBOPUHfDha36M:b2c0zcb)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @xhcao)

File: tfjs-backend-webgpu/src/matmul_splitK_webgpu.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    const tileBOuter = this.workGroupSize[0] * this.elementsPerThread[0];
+    const rowPerThread = this.elementsPerThread[1];
+    const colPerThread = this.elementsPerThread[0];
+    const colPerThreadA = this.tileInner / this.workGroupSize[0];
+    const rowPerThreadB = this.tileInner / this.workGroupSize[1];
+    util.assert(
+        this.tileInner % this.workGroupSize[0] === 0 &&
+            this.tileInner % this.workGroupSize[1] === 0,
+        () =>
+            `tileInner ${this.tileInner} must be divisible by workGroupSize[0]${
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/matmul_splitK_webgpu.ts:155]: Is it better to show `tileInner` must larger or equal than workgroupsize[0]|[1] directly?

**Comment by xhcao** [comment]
Two other questions,
1. Could list the performance time of `BiasActivationProgram` here, because it is also a part of additions.
2. Because we split a kernel into several kernels, will it add the cpu time here?

**Comment by qjia7** [comment]
> 1. Could list the performance time of `BiasActivationProgram` here, because it is also a part of additions.

The data I listed comes from `bodypix-ResNet50-tensor-0.5` which doesn't have a fused conv2d. So there is no additional `BiasActivationProgram`. But I suppose it's very lightweight (less than 0.05 ms) since it's only applied to the output, which is a small size matrix.

> 2. Because we split a kernel into several kernels, will it add the cpu time here?

Unless it happens that gpu waits cpu. Otherwise it's not a problem. And I assume that `gpu waits cpu` only happen when it's a very lightweight model. For such kind of model, maybe wasm will be the recommended backend.

**Review by xhcao — APPROVED** [review]
Jiajia, thank you for reply. This algorithm is very good to use the GPU resources.

**Pull request merged**